pkg/certificate/daemon.go

```diff
@@ -283,105 +279,97 @@ func Certificated(idConfig *config.IdentityConfig, stopChan <-chan struct{}) (er
+			log.Infof("Will refresh key[%s], cert[%s] and certificates for roles[%v] with provider[%s], backup[%s] and secret[%s] within %s", cs.idConfig.KeyFile, cs.idConfig.CertFile, cs.idConfig.TargetDomainRoles, cs.idConfig.ProviderService, cs.idConfig.Backup, cs.idConfig.CertSecret, cs.idConfig.Refresh)
 
 			select {
+			case <-cs.shutdownChan:
+				log.Info("Stopped certificate provider daemon")
+				return
 			case <-t.C:
-				err := backoff.RetryNotify(run, getExponentialBackoff(), notifyOnErr)
-				if err != nil {
-					log.Errorf("Failed to refresh x509 certificate after multiple retries: %s", err.Error())
+				// skip refresh if context is done but Shutdown() is not called
+				if ctx.Err() != nil {
+					log.Infof("Skipped to refresh key[%s], cert[%s] and certificates for roles[%v] with provider[%s], backup[%s] and secret[%s]", cs.idConfig.KeyFile, cs.idConfig.CertFile, cs.idConfig.TargetDomainRoles, cs.idConfig.ProviderService, cs.idConfig.Backup, cs.idConfig.CertSecret)
+					continue
 				}
-			case <-stopChan:
-				log.Info("Initiating shutdown of certificate provider daemon ...")
-				err = deleteRequest()
+
+				// backoff retry until REFRESH_INTERVAL / 4 OR context is done
+				err := backoff.RetryNotify(cs.run, newExponentialBackOff(ctx, cs.idConfig.Refresh/4), notifyOnErr)
 				if err != nil {
-					log.Errorf("Failed to delete x509 certificate Instance ID record: %s", err.Error())
-				}
-				close(healthcheckChan)
-				close(metricsChan)
-				close(tokenChan)
-				if tokenSdChan != nil {
-					<-tokenSdChan
+					log.Errorf("Failed to refresh certificates after multiple retries: %s", err.Error())
 				}
-				if metricsSdChan != nil {
-					<-metricsSdChan
-				}
-				if healthcheckSdChan != nil {
-					<-healthcheckSdChan
-				}
-				return
 			}
 		}
 	}()
 
-	return nil, shutdownChan
+	return nil
+}
+
+func (cs *certService) Shutdown() {
+	log.Info("Initiating shutdown of certificate provider daemon ...")
+	close(cs.shutdownChan)
+
+	// wait for graceful shutdown
+	cs.shutdownWg.Wait()
+
+	// delete x509 certificate record to prevent future refresh
+	// P.S. Shutdown() will ONLY run on mode=refresh, no need to check for idConfig.Init
+	if cs.idConfig.DeleteInstanceID && cs.handler.InstanceID() != "" {
+		log.Infoln("Attempting to delete x509 certificate record from identity provider...")
+		err := cs.handler.DeleteX509CertRecord()
+		if err != nil {
+			log.Warnf("Failed to delete x509 certificate Instance ID record: %s", err.Error())
+		} else {
+			log.Infof("Successfully deleted x509 certificate Instance ID record[%s]", cs.handler.InstanceID())
+		}
+	}
+}
+
+// newExponentialBackOff returns a backoff config with first retry delay of 5s. Allow cancel by context.
+func newExponentialBackOff(ctx context.Context, maxElapsedTime time.Duration) backoff.BackOff {
```

Comment:
@WindzCUHK I don't believe it's necessary to change this function name from `getExponentialBackoff` since we're just replacing it.

Reply:
new() creates an object. get() get an existing object.
Using new() as function name is more suitable.